compiler/src/dotty/tools/dotc/core/NullOpsDecorator.scala

```diff
@@ -0,0 +1,147 @@
+package dotty.tools.dotc.core
+
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Symbols.defn
+import dotty.tools.dotc.core.Types.{AndType, ClassInfo, ConstantType, OrType, Type, TypeBounds, TypeMap, TypeProxy}
+
+/** Defines operations on nullable types. */
+object NullOpsDecorator {
+
+  implicit class NullOps(val self: Type) {
+    /** Is this type a reference to `Null`, possibly after aliasing? */
+    def isNullType(implicit ctx: Context): Boolean = self.isRef(defn.NullClass)
+
+    /** Is self (after widening and dealiasing) a type of the form `T | Null`? */
+    def isNullableUnion(implicit ctx: Context): Boolean = {
+      assert(ctx.explicitNulls)
+      self.widenDealias.normNullableUnion match {
+        case OrType(_, right) => right.isNullType
+        case _ => false
+      }
+    }
+
+    /** Is this type exactly `JavaNull` (no vars, aliases, refinements etc allowed)? */
+    def isJavaNullType(implicit ctx: Context): Boolean = {
+      assert(ctx.explicitNulls)
+      // We can't do `self == defn.JavaNull` because when trees are unpickled new references
+      // to `JavaNull` could be created that are different from `defn.JavaNull`.
+      // Instead, we compare the symbol.
+      self.isDirectRef(defn.JavaNullAlias)
+    }
+
+    /** Is self (after widening and dealiasing) a type of the form `T | JavaNull`? */
+    def isJavaNullableUnion(implicit ctx: Context): Boolean = {
+      assert(ctx.explicitNulls)
+      def peelOff(tpe: Type): Boolean = tpe match {
+        case OrType(left, right) =>
+          right.isJavaNullType || right.isNullType && peelOff(left)
+        case _ => false
+      }
+      val norm = self.widenDealias.normNullableUnion
+      // We can't just look at the r.h.s of the normalized union.
+      // The problem is that after normalizing we could get a type like `(String | JavaNull) | Null`,
+      // where the r.h.s of the union isn't a `JavaNull`, but if we keep peeling nulls off starting
+      // from the right we'll eventually get to the `JavaNull`.
+      peelOff(norm)
+    }
+
+    /** Is this type guaranteed not to have `null` as a value? */
+    final def isNotNull(implicit ctx: Context): Boolean = self match {
+      case tp: ConstantType => tp.value.value != null
+      case tp: ClassInfo => !tp.cls.isNullableClass && tp.cls != defn.NothingClass
+      case tp: TypeBounds => tp.lo.isNotNull
+      case tp: TypeProxy => tp.underlying.isNotNull
+      case AndType(tp1, tp2) => tp1.isNotNull || tp2.isNotNull
+      case OrType(tp1, tp2) => tp1.isNotNull && tp2.isNotNull
+      case _ => false
+    }
+
+    /** Normalizes unions so that all `Null`s (or aliases to `Null`) appear to the right of all other types.
+     *  In the process, it also flattens the type so that there are no nested unions at the top level.
+     *  e.g. `Null | (T1 | Null) | T2` => `T1 | T2 | Null | Null`
+     */
+    def normNullableUnion(implicit ctx: Context): Type = {
```

Comment:
Should be reformulated to be more efficient, since it is called in many hotspots. First rule to get there: Avoid object allocations of all forms!

Reply:
Done